contracts/savings/SavingsManager.sol

```diff
@@ -136,58 +138,96 @@ contract SavingsManager is ISavingsManager, PausableModule {
-                // e.g. (4.1667e12 * 1e18) / 5.7..e13 = 7.1e16 or 7.1% apr
-                uint256 extrapolatedAPY = percentageIncrease.divPrecisely(yearsSinceLastCollection);
-
-                require(extrapolatedAPY < MAX_APY, "Interest protected from inflating past maxAPY");
-
-                emit InterestCollected(_mAsset, interestCollected, totalSupply, extrapolatedAPY);
-
-                // 4. Distribute the interest
-                // Calculate the share for savers (95e16 or 95%)
-                uint256 saversShare = interestCollected.mulTruncate(savingsRate);
-
-                // Call depositInterest on contract
-                savingsContract.depositInterest(saversShare);
-
-                emit InterestDistributed(_mAsset, saversShare);
-            } else {
-                emit InterestCollected(_mAsset, 0, totalSupply, 0);
-            }
+            emit InterestDistributed(_mAsset, saversShare);
         } else {
-            emit WaitFor(THIRTY_MINUTES.sub(timeSinceLastCollection));
+            emit InterestCollected(_mAsset, 0, totalSupply, 0);
+        }
+    }
+
+    /**
+     * @dev Validates that an interest collection does not exceed a maximum APY. If last collection
+     * was under 30 mins ago, simply check it does not exceed 10bps
+     * @param _newSupply               New total supply of the mAsset
+     * @param _interest                Increase in total supply since last collection
+     * @param _timeSinceLastCollection Seconds since last collection
+     */
+    function _validateCollection(uint256 _newSupply, uint256 _interest, uint256 _timeSinceLastCollection)
+        internal
+        pure
+        returns (uint256 extrapolatedAPY)
+    {
+        // e.g. day: (86400 * 1e18) / 3.154e7 = 2.74..e15
+        // e.g. 30 mins: (1800 * 1e18) / 3.154e7 = 5.7..e13
+        // e.g. epoch: (1593596907 * 1e18) / 3.154e7 = 50.4..e18
+        uint256 yearsSinceLastCollection =
+            _timeSinceLastCollection.divPrecisely(SECONDS_IN_YEAR);
+
+        // Percentage increase in total supply
+        // e.g. (1e20 * 1e18) / 1e24 = 1e14 (or a 0.01% increase)
+        // e.g. (5e18 * 1e18) / 1.2e24 = 4.1667e12
+        // e.g. (1e19 * 1e18) / 1e21 = 1e16
+        uint256 oldSupply = _newSupply.sub(_interest);
+        uint256 percentageIncrease = _interest.divPrecisely(oldSupply);
+
+        // e.g. 0.01% (1e14 * 1e18) / 2.74..e15 = 3.65e16 or 3.65% apr
+        // e.g. (4.1667e12 * 1e18) / 5.7..e13 = 7.1e16 or 7.1% apr
+        // e.g. (1e16 * 1e18) / 50e18 = 2e14
+        extrapolatedAPY = percentageIncrease.divPrecisely(yearsSinceLastCollection);
+
+        //      If over 30 mins, extrapolate APY
+        if(_timeSinceLastCollection > THIRTY_MINUTES) {
+            require(extrapolatedAPY < MAX_APY, "Interest protected from inflating past maxAPY");
+        } else {
+            require(percentageIncrease < TEN_BPS, "Interest protected from inflating past 10 Bps");
```

Comment:
It seems the goal is to make sure the supply isn't inflated by 0.1% or more if the time between collections is 30 minutes or less.

However, the current implementation only checks this atomically per collection. It doesn't consider the past inflation already experienced in the last _X_ minutes.

**Example:** Bob decides he wants to corrupt mStable's asset through inflation. He exerts massive trading volume. To bypass the `MAX_APY` check, he makes sure he's updating the `timeSinceLastCollection` within 30 minutes between each collection. 
Now, he makes sure he's only inflating the supply by 0.09%, then updating the `timeSinceLastCollection` (through calling the `collectAndDistributeInterest()`). He can now inflate the asset by over 2.7% in 30 minutes (assuming he does this every minute), and bypass any restrictive mechanisms on inflation that are set.

That example uses a malicious actor, but there's natural ways this could happen too. The point being, if I'm understanding the reasoning of the restriction, the implementation should consider the past inflation too when it's been 30 minutes or less since last collection.

A state variable could be added that get's incremented by the inflation experienced as long as it's in one 30 minute collection time. Once the collection time overflows 30 minutes, it can be zeroed out and restarted. This state variable is what would be used in the above check.

`require(periodPercentageIncrease < TEN_BPS, "...");`

Where the `period...` variable is referring to the sum of inflation in the current 30 minute collection period.

_Note, copying my original comment over from the private dev. repo_

Reply:
This comment is outdated due to the summing of the period yield in line 168